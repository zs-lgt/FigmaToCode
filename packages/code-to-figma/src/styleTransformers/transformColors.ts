import { Color } from '../types';
import * as _parse from 'color-parse';
import gradientParse from '../helpers/gradientParser';

const parse = typeof _parse === 'function' ? _parse : _parse.default;

export const colorToRGB = (color: Color): RGB => {
    const parsed = parse(color);
    if (parsed && parsed.space === 'rgb') {
        return {
            r: parsed.values[0] / 255.0,
            g: parsed.values[1] / 255.0,
            b: parsed.values[2] / 255.0
        };
    }
};

export const colorToRGBA = (color: Color, opacity?: number): RGBA => {
    const parsed = parse(color);
    if (parsed && parsed.space === 'rgb') {
        return {
            r: parsed.values[0] / 255.0,
            g: parsed.values[1] / 255.0,
            b: parsed.values[2] / 255.0,
            a: typeof opacity === 'number' ? opacity : parsed.alpha
        };
    }
};

const transformAngleToFigmaTransformation = (value): Transform => {
    const angle = value - Math.PI / 2;

    // **
    // There is rotation relative to center through affine transformation
    // (xc, yc) is a center of the imaginary 1x1 quad
    // (dxc, dyc) is a delta between the old center and the new center (after transformation)
    // Transformation matrix has transform to the delta value, because center position should not change.
    // **
    const xc = 0.5;
    const yc = 0.5;
    const dxc = xc - (Math.cos(angle) * xc + Math.sin(angle) * yc);
    const dyc = yc - (-Math.sin(angle) * xc + Math.cos(angle) * yc);
    return [[Math.cos(angle), Math.sin(angle), dxc], [-Math.sin(angle), Math.cos(angle), dyc]];
};

export const colorToPaint = (color: Color): Paint => {
    let parsedGradients;
    try {
        parsedGradients = gradientParse.parse(color);
    } catch (e) {}

    const parsedGradient = parsedGradients && parsedGradients[0];
    if (parsedGradient && parsedGradient.type === 'linear-gradient') {
        let gradientTransform = transformAngleToFigmaTransformation(Math.PI);
        if (parsedGradient.orientation) {
            const orientation = parsedGradient.orientation;
            if (orientation.type === 'directional') {
                if (orientation.value === 'right') {
                    gradientTransform = transformAngleToFigmaTransformation(Math.PI / 2);
                } else if (orientation.value === 'left') {
                    gradientTransform = transformAngleToFigmaTransformation(-Math.PI / 2);
                } else if (orientation.value === 'top') {
                    gradientTransform = transformAngleToFigmaTransformation(0);
                }
            } else if (orientation.type == 'angular') {
                const angle = (parseFloat(orientation.value) / 180) * Math.PI;
                gradientTransform = transformAngleToFigmaTransformation(angle);
            }
        }

        return {
            type: 'GRADIENT_LINEAR',
            gradientTransform,
            gradientStops: parsedGradient.colorStops.map(({ type, value, length }, index) => ({
                position: length && length.type === '%' ? parseFloat(length.value) / 100 : index === 0 ? 0 : 1,
                color: colorToRGBA(type === 'hex' ? `#${value}` : value)
            }))
        };
    }
    const { a: alpha, ...rbg } = colorToRGBA(color);
    return { type: 'SOLID', color: rbg, opacity: alpha };
};
