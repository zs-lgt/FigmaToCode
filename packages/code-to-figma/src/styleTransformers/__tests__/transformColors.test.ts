import { colorToPaint, colorToRGB, colorToRGBA } from '../transformColors';

describe('transformColors', () => {
    it('colorToRGB #9b9fff', () => {
        const result = colorToRGB('#9b9fff');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB #ff0000', () => {
        const result = colorToRGB('#ff0000');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGBA #ff0000 0.5', () => {
        const result = colorToRGBA('#ff0000', 0.5);
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB red', () => {
        const result = colorToRGB('red');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB aliceblue', () => {
        const result = colorToRGB('aliceblue');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB rgb(255, 0, 255)', () => {
        const result = colorToRGB('rgb(255, 0, 255)');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB rgba(255, 255, 255, 1.0)', () => {
        const result = colorToRGB('rgba(255, 255, 255, 1.0)');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB #ff00ff00', () => {
        const result = colorToRGB('#ff00ff00');
        expect(result).toMatchSnapshot();
    });

    it('colorToRGB transparent', () => {
        const result = colorToRGB('transparent');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint #9b9fff', () => {
        const result = colorToPaint('#9b9fff');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint rgba(27,31,35,0.2)', () => {
        const result = colorToPaint('rgba(27,31,35,0.2)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint linear-gradient(#e66465, #9198e5)', () => {
        const result = colorToPaint('linear-gradient(#e66465, #9198e5)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint linear-gradient(to left, #e66465, #9198e5)', () => {
        const result = colorToPaint('linear-gradient(to left, #e66465, #9198e5)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint linear-gradient(to right, #e66465, #9198e5)', () => {
        const result = colorToPaint('linear-gradient(to right, #e66465, #9198e5)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint linear-gradient(to top, #e66465, #9198e5)', () => {
        const result = colorToPaint('linear-gradient(to top, #e66465, #9198e5)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint linear-gradient(to bottom, #e66465, #9198e5)', () => {
        const result = colorToPaint('linear-gradient(to bottom, #e66465, #9198e5)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint linear-gradient(30deg, #e66465, #9198e5)', () => {
        const result = colorToPaint('linear-gradient(30deg, #e66465, #9198e5)');
        expect(result).toMatchSnapshot();
    });

    it('colorToPaint with percentage: linear-gradient(#e66465, #9198e5 80%)', () => {
        const result = colorToPaint('linear-gradient(#e66465, #9198e5 80%)');
        expect(result).toMatchSnapshot();
    });
});
