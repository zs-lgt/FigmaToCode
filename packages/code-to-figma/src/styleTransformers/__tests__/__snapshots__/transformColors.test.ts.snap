// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformColors colorToPaint #9b9fff 1`] = `
Object {
  "color": Object {
    "b": 1,
    "g": 0.6235294117647059,
    "r": 0.6078431372549019,
  },
  "opacity": 1,
  "type": "SOLID",
}
`;

exports[`transformColors colorToPaint linear-gradient(#e66465, #9198e5) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 1,
    },
  ],
  "gradientTransform": Array [
    Array [
      6.123233995736766e-17,
      1,
      0,
    ],
    Array [
      -1,
      6.123233995736766e-17,
      1,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToPaint linear-gradient(30deg, #e66465, #9198e5) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 1,
    },
  ],
  "gradientTransform": Array [
    Array [
      0.4999999999999999,
      -0.8660254037844387,
      0.6830127018922194,
    ],
    Array [
      0.8660254037844387,
      0.4999999999999999,
      -0.1830127018922193,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToPaint linear-gradient(to bottom, #e66465, #9198e5) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 1,
    },
  ],
  "gradientTransform": Array [
    Array [
      6.123233995736766e-17,
      1,
      0,
    ],
    Array [
      -1,
      6.123233995736766e-17,
      1,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToPaint linear-gradient(to left, #e66465, #9198e5) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 1,
    },
  ],
  "gradientTransform": Array [
    Array [
      -1,
      -1.2246467991473532e-16,
      1,
    ],
    Array [
      1.2246467991473532e-16,
      -1,
      1,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToPaint linear-gradient(to right, #e66465, #9198e5) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 1,
    },
  ],
  "gradientTransform": Array [
    Array [
      1,
      0,
      0,
    ],
    Array [
      -0,
      1,
      0,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToPaint linear-gradient(to top, #e66465, #9198e5) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 1,
    },
  ],
  "gradientTransform": Array [
    Array [
      6.123233995736766e-17,
      -1,
      1,
    ],
    Array [
      1,
      6.123233995736766e-17,
      0,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToPaint rgba(27,31,35,0.2) 1`] = `
Object {
  "color": Object {
    "b": 0.13725490196078433,
    "g": 0.12156862745098039,
    "r": 0.10588235294117647,
  },
  "opacity": 0.2,
  "type": "SOLID",
}
`;

exports[`transformColors colorToPaint with percentage: linear-gradient(#e66465, #9198e5 80%) 1`] = `
Object {
  "gradientStops": Array [
    Object {
      "color": Object {
        "a": 1,
        "b": 0.396078431372549,
        "g": 0.39215686274509803,
        "r": 0.9019607843137255,
      },
      "position": 0,
    },
    Object {
      "color": Object {
        "a": 1,
        "b": 0.8980392156862745,
        "g": 0.596078431372549,
        "r": 0.5686274509803921,
      },
      "position": 0.8,
    },
  ],
  "gradientTransform": Array [
    Array [
      6.123233995736766e-17,
      1,
      0,
    ],
    Array [
      -1,
      6.123233995736766e-17,
      1,
    ],
  ],
  "type": "GRADIENT_LINEAR",
}
`;

exports[`transformColors colorToRGB #9b9fff 1`] = `
Object {
  "b": 1,
  "g": 0.6235294117647059,
  "r": 0.6078431372549019,
}
`;

exports[`transformColors colorToRGB #ff0000 1`] = `
Object {
  "b": 0,
  "g": 0,
  "r": 1,
}
`;

exports[`transformColors colorToRGB #ff00ff00 1`] = `
Object {
  "b": 1,
  "g": 0,
  "r": 1,
}
`;

exports[`transformColors colorToRGB aliceblue 1`] = `
Object {
  "b": 1,
  "g": 0.9725490196078431,
  "r": 0.9411764705882353,
}
`;

exports[`transformColors colorToRGB red 1`] = `
Object {
  "b": 0,
  "g": 0,
  "r": 1,
}
`;

exports[`transformColors colorToRGB rgb(255, 0, 255) 1`] = `
Object {
  "b": 1,
  "g": 0,
  "r": 1,
}
`;

exports[`transformColors colorToRGB rgba(255, 255, 255, 1.0) 1`] = `
Object {
  "b": 1,
  "g": 1,
  "r": 1,
}
`;

exports[`transformColors colorToRGB transparent 1`] = `
Object {
  "b": 0,
  "g": 0,
  "r": 0,
}
`;

exports[`transformColors colorToRGBA #ff0000 0.5 1`] = `
Object {
  "a": 0.5,
  "b": 0,
  "g": 0,
  "r": 1,
}
`;
